File: Sources/Runtime/Components/MelodyProgress.swift
import SwiftUI
import Core

/// Renders a determinate or indeterminate progress indicator with an optional label.
struct MelodyProgress: View {
    let definition: ComponentDefinition
    let resolvedValue: String?
    let resolvedLabel: String?

    var body: some View {
        Group {
            if let value = progressValue {
                if let label = resolvedLabel, !label.isEmpty {
                    ProgressView(label, value: value)
                } else {
                    ProgressView(value: value)
                }
            } else {
                if let label = resolvedLabel, !label.isEmpty {
                    ProgressView(label)
                } else {
                    ProgressView()
                }
            }
        }
        .melodyStyle(definition.style)
    }

    private var progressValue: Double? {
        guard let str = resolvedValue, !str.isEmpty else { return nil }
        return Double(str)
    }
}

File: Android/melody-runtime/src/main/java/com/melody/runtime/components/MelodyProgress.kt
package com.melody.runtime.components

import androidx.compose.foundation.layout.Column
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import com.melody.core.schema.ComponentDefinition

@Composable
fun MelodyProgress(
    definition: ComponentDefinition,
    resolvedValue: String?,
    resolvedLabel: String?
) {
    val progressValue = resolvedValue?.toDoubleOrNull()
    val modifier = Modifier.melodyStyle(definition.style)
    val hasLabel = !resolvedLabel.isNullOrEmpty()

    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        modifier = modifier
    ) {
        if (hasLabel) {
            Text(resolvedLabel!!)
        }
        if (progressValue != null) {
            LinearProgressIndicator(
                progress = { progressValue.toFloat().coerceIn(0f, 1f) }
            )
        } else {
            CircularProgressIndicator()
        }
    }
}
